config/config_helpers.go

```diff
@@ -327,6 +328,81 @@ func getAWSCallerIdentityUserID(include *IncludeConfig, terragruntOptions *optio
 	return "", err
 }
 
+// Parse the terragrunt config and return a representation that can be used as a reference. If given a default value,
+// this will return the default if the terragrunt config file does not exist.
+func readTerragruntConfig(configPath string, defaultVal *cty.Value, terragruntOptions *options.TerragruntOptions) (cty.Value, error) {
+	// target config check: make sure the target config exists. If the file does not exist, and there is no default val,
+	// return an error. If the file does not exist but there is a default val, return the default val. Otherwise,
+	// proceed to parse the file as a terragrunt config file.
+	targetConfig := getCleanedTargetConfigPath(configPath, terragruntOptions)
+	targetConfigFileExists := util.FileExists(targetConfig)
+	if !targetConfigFileExists && defaultVal == nil {
+		return cty.NilVal, errors.WithStackTrace(TerragruntConfigNotFound{Path: targetConfig})
+	} else if !targetConfigFileExists {
+		return *defaultVal, nil
+	}
+
+	// We update the context of terragruntOptions to the config being read in.
+	targetOptions := terragruntOptions.Clone(targetConfig)
+	config, err := ParseConfigFile(configPath, targetOptions, nil)
+	if err != nil {
+		return cty.NilVal, err
+	}
+	// We have to set the rendered outputs here because ParseConfigFile will not do so on the TerragruntConfig. The
+	// outputs are stored in a special map that is used only for rendering and thus is not available when we try to
+	// serialize the config for consumption.
+	// NOTE: this will not call terragrunt output, since all the values are cached from the ParseConfigFile call
+	// NOTE: we don't use range here because range will copy the slice, thereby undoing the set attribute.
+	for i := 0; i < len(config.TerragruntDependencies); i++ {
+		config.TerragruntDependencies[i].setRenderedOutputs(targetOptions)
```

Comment:
Quick sanity check: any time I see a mutable call like this, I have to ask if it is thread / goroutine safe? E.g., Does this work properly during `apply-all`?

Reply:
Yes, because the `config` object is function scoped.